binding.c

```diff
@@ -135,94 +194,113 @@ on_udx_message (udx_socket_t *self, ssize_t read_len, const uv_buf_t *buf, const
 
   memcpy(n->udx->read_buf, buf->base, buf->len);
 
-  napi_env env = n->env;
+  js_env_t *env = n->env;
+
+  js_handle_scope_t *scope;
+  err = js_open_handle_scope(env, &scope);
+  assert(err == 0);
 
-  napi_handle_scope scope;
-  napi_open_handle_scope(env, &scope);
+  js_value_t *ctx;
+  err = js_get_reference_value(env, n->ctx, &ctx);
+  assert(err == 0);
 
-  napi_value ctx;
-  napi_get_reference_value(env, n->ctx, &ctx);
+  js_value_t *callback;
+  err = js_get_reference_value(env, n->on_message, &callback);
+  assert(err == 0);
 
-  napi_value callback;
-  napi_get_reference_value(env, n->on_message, &callback);
+  js_value_t *argv[4];
+  err = js_create_uint32(env, read_len, &(argv[0]));
+  assert(err == 0);
+  err = js_create_uint32(env, port, &(argv[1]));
+  assert(err == 0);
+  err = js_create_string_utf8(env, (utf8_t *) ip, JS_AUTO_LENGTH, &(argv[2]));
+  assert(err == 0);
+  err = js_create_uint32(env, family, &(argv[3]));
+  assert(err == 0);
 
-  napi_value argv[4];
-  napi_create_uint32(env, read_len, &(argv[0]));
-  napi_create_uint32(env, port, &(argv[1]));
-  napi_create_string_utf8(env, ip, NAPI_AUTO_LENGTH, &(argv[2]));
-  napi_create_uint32(env, family, &(argv[3]));
+  js_value_t *res;
 
-  napi_value res;
+  err = js_call_function(env, ctx, callback, 4, argv, &res);
```

Comment:
This should do a microtask checkpoint assuming that `on_udx_message()` is an I/O callback.